x-pack/legacy/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -117,18 +123,50 @@ export class KerberosAuthenticationProvider extends BaseAuthenticationProvider {
     const [, kerberosTicket] = request.headers.authorization.split(/\s+/);
 
     // First attempt to exchange SPNEGO token for an access token.
-    let tokens: { access_token: string; refresh_token: string };
+    let tokens: {
+      access_token: string;
+      refresh_token: string;
+      kerberos_authentication_response_token?: string;
+    };
     try {
       tokens = await this.options.client.callWithInternalUser('shield.getAccessToken', {
         body: { grant_type: '_kerberos', kerberos_ticket: kerberosTicket },
       });
     } catch (err) {
       this.debug(`Failed to exchange SPNEGO token for an access token: ${err.message}`);
+
+      // Check if SPNEGO context wasn't established and we have a response token to return to the client.
+      const challenge =
+        getErrorStatusCode(err) === 401 ? this.getNegotiateChallenge(err) : undefined;
+      if (challenge) {
+        const [, responseToken] = challenge.split(/\s+/);
+        if (responseToken) {
```

Comment:
> What if we don't have a responseToken here?

That's a good point! initially I was thinking that we should never get 401 here with `Negotiate` only (e.g. in case of replay attack I expected to see here something different than `401 WWW-Authenticate: Negotiate`), but I've just checked and there are indeed valid cases when we may get `Negotiate` without token here. So I'll change that to something like this and write an integration test (in progress):
```ts
return AuthenticationResult.failed(Boom.unauthorized(), {
  authResponseHeaders: {
    [WWWAuthenticateHeaderName]: responseToken ? `Negotiate ${responseToken}` : 'Negotiate',
  },
});
```

> Or, what if challenge.split(/\s+/).length > 2? It looks like we're relying on getting a well-formed challenge back from Elasticsearch, but we might want to handle (or at least log) cases where we expected a token, but didn't get one. Proxies in between have been known to mangle responses.

Yeah, it won't hurt if I log such things with `warn`, thanks!

Reply:
Fixed, pushed